src/pkg/systemartifact/manager.go

```diff
@@ -0,0 +1,231 @@
+
+	// Cleanup cleans up the system artifacts (tracking records as well as blobs) based on the
+	// artifact records selected by the CleanupCriteria registered for each vendor type.
+	// Returns the total number of records deleted, the reclaimed size and any error (if encountered)
+	Cleanup(ctx context.Context) (int64, int64, error)
+}
+
+type systemArtifactManager struct {
+	regCli                  registry.Client
+	dao                     dao.DAO
+	defaultCleanupCriterion CleanupCriteria
+	cleanupCriteria         map[string]CleanupCriteria
+}
+
+func NewManager() Manager {
+	sysArtifactMgr := &systemArtifactManager{
+		regCli:                  registry.Cli,
+		dao:                     dao.NewSystemArtifactDao(),
+		defaultCleanupCriterion: DefaultCleanupCriteria,
+		cleanupCriteria:         make(map[string]CleanupCriteria),
+	}
+	return sysArtifactMgr
+}
+
+func (mgr *systemArtifactManager) Create(ctx context.Context, artifactRecord *model.SystemArtifact, reader io.Reader) (int64, error) {
+
+	id, err := mgr.dao.Create(ctx, artifactRecord)
+
+	if err != nil {
+		return int64(0), err
+	}
+
+	repoName := mgr.getRepositoryName(artifactRecord.Vendor, artifactRecord.Repository)
+	err = mgr.regCli.PushBlob(repoName, artifactRecord.Digest, artifactRecord.Size, reader)
+
+	if err != nil {
+
+		mgr.dao.Delete(ctx, artifactRecord.Vendor, artifactRecord.Repository, artifactRecord.Digest)
+		return int64(0), err
+	}
+	return id, nil
+}
+
+func (mgr *systemArtifactManager) Read(ctx context.Context, vendor string, repository string, digest string) (io.ReadCloser, error) {
+	sa, err := mgr.dao.Get(ctx, vendor, repository, digest)
+	if err != nil {
+		return nil, err
+	}
+	repoName := mgr.getRepositoryName(vendor, repository)
+	_, readCloser, err := mgr.regCli.PullBlob(repoName, sa.Digest)
+	if err != nil {
+		return nil, err
+	}
+	return readCloser, nil
+}
+
+func (mgr *systemArtifactManager) Delete(ctx context.Context, vendor string, repository string, digest string) error {
+
+	repoName := mgr.getRepositoryName(vendor, repository)
+	err := mgr.regCli.DeleteBlob(repoName, digest)
+
+	if err != nil {
+		return err
+	}
+
+	err = mgr.dao.Delete(ctx, vendor, repository, digest)
+
+	return err
+}
+
+func (mgr *systemArtifactManager) Exists(ctx context.Context, vendor string, repository string, digest string) (bool, error) {
+	_, err := mgr.dao.Get(ctx, vendor, repository, digest)
+	if err != nil {
+		return false, err
+	}
+
+	repoName := mgr.getRepositoryName(vendor, repository)
+	exist, err := mgr.regCli.BlobExist(repoName, digest)
+
+	if err != nil {
+		return false, err
+	}
+
+	return exist, nil
+}
+
+func (mgr *systemArtifactManager) GetStorageSize(ctx context.Context) (int64, error) {
+	listQuery := q.Query{}
+	artifacts, err := mgr.dao.List(ctx, &listQuery)
+
+	if err != nil {
+		return int64(0), err
+	}
+
+	size := int64(0)
+
+	for _, artifact := range artifacts {
+		size += artifact.Size
```

Comment:
We need to consider potential performance issue if it has a lot of artifacts, `List` will select * from database but we only need size column, maybe could use SUM SQL to aggregate result?